include/sparrow/layout/union_layout/union_array.hpp

```diff
@@ -0,0 +1,201 @@
+        using union_array_crtp_base<sparse_union_array>::union_array_crtp_base;
+    private:
+        std::size_t element_offset(std::size_t i) const;
+        friend class union_array_crtp_base<sparse_union_array>;
+    };
+
+    template <class DERIVED>
+    union_array_crtp_base<DERIVED>::union_array_crtp_base(arrow_proxy proxy)
+    :   m_proxy(std::move(proxy)),
+        p_type_ids(reinterpret_cast<std::uint8_t*>(m_proxy.buffers()[0].data())),
+        m_children(m_proxy.children().size(), nullptr),
+        m_type_id_map(parse_type_id_map(m_proxy.format()))
+    {
+        for (std::size_t i = 0; i < m_children.size(); ++i)
+        {
+            m_children[i] = array_factory(m_proxy.children()[i].view());
+        }
+    }
+
+    template <class DERIVED>
+    auto union_array_crtp_base<DERIVED>::operator[](std::size_t i) const -> value_type
+    {   
+        const auto type_id = static_cast<std::size_t>(p_type_ids[i]);
+        const auto child_index = m_type_id_map[type_id];
+        const auto offset = this->derived_cast().element_offset(i);
+        return array_element(*m_children[child_index], static_cast<std::size_t>(offset));
+    }
+
+    template <class DERIVED>
+    auto union_array_crtp_base<DERIVED>::operator[](std::size_t i) -> value_type
+    {   
+        return static_cast<const derived_type&>(*this)[i];
+    }
+
+    template <class DERIVED>
+    std::size_t union_array_crtp_base<DERIVED>::size() const
+    {
+        return m_proxy.length();
+    }
+
+    template <class DERIVED>
+    auto union_array_crtp_base<DERIVED>::begin() -> iterator
+    {
+        return iterator(detail::layout_bracket_functor<derived_type, value_type>{this}, 0);
+    }
+
+    template <class DERIVED>
+    auto union_array_crtp_base<DERIVED>::end() -> iterator
+    {
+        return iterator(detail::layout_bracket_functor<derived_type, value_type>{this}, this->size());
+    }
+
+    template <class DERIVED>
+    auto union_array_crtp_base<DERIVED>::begin() const -> const_iterator    
+    {
+        return cbegin();
+    }
+
+    template <class DERIVED>
+    auto union_array_crtp_base<DERIVED>::end() const -> const_iterator   
+    {
+        return cend();
+    }
+
+    template <class DERIVED>
+    auto union_array_crtp_base<DERIVED>::cbegin() const -> const_iterator   
+    {
+        return const_iterator(detail::layout_bracket_functor<const derived_type, value_type>{this}, 0);
+    }
+
+    template <class DERIVED>
+    auto union_array_crtp_base<DERIVED>::cend() const  -> const_iterator
+    {
+        return const_iterator(detail::layout_bracket_functor<const derived_type, value_type>{this}, this->size());
+    }
+
+    #ifdef __GNUC__
+    #    pragma GCC diagnostic push
+    #    pragma GCC diagnostic ignored "-Wcast-align"
+    #endif
+    inline dense_union_array::dense_union_array(arrow_proxy proxy)
+    :   union_array_crtp_base(std::move(proxy)),
+        p_offsets(reinterpret_cast<std::int32_t*>(m_proxy.buffers()[1].data()))
```

Comment:
Can you replace the "1" magic number by a variable

Reply:
are you also fine with:
```C++
 p_offsets(reinterpret_cast<std::int32_t*>(m_proxy.buffers()[1/*index of offsets*/].data()))
```